### DISCUSSION TITLE
Efficient Implementation

### DISCUSSION THREAD

**Discussion by christian-heise** [chesskit-app/chesskit-engine#5] [Q&A]
Hello, I implemented the chesskit-engine into my in-production app and it works great. However the workload for the iPhone gets really high and it therefor gets hot.

Is there maybe a way to implement the engine more efficiently? My implementation is the following:

```
func getEngineMoves() {
        let position = self.game.position
        let fen = FenSerialization.default.serialize(position: position)

        DispatchQueue.global(qos: .userInitiated).async {
            self.engine.send(command: .position(.fen(fen)))
            self.engine.send(command: .go(depth: 20))
            self.engine.receiveResponse = { response in
                DispatchQueue.main.async {
                    switch response {
                    case let .info(info):
                        if let score = info.score?.cp {
                            self.evaluation = score
                        }
                        if let mateInXMoves = info.score?.mate {
                            self.mateInXMoves = mateInXMoves
                        }
                    default:
                        break
                    }
                }
            }
        }
    }
```

The `self.engine.send(command: .stop)` gets called before the function is called. The function is called every time the user makes a move.

Or maybe it's possible to throttle the performance of the engine? Limit the threads or something.
Edit: I actually found out how to limit the threads. However, are there recommendations on how many to use?

Thanks for the great package!

**Comment by pdil** [comment]
The number of threads is limited by the CPU cores on the system the engine is running on. It's up to you how many threads to use depending on your needs. More threads achieves higher search depth faster but at the cost of more processing power, battery life, etc.

Another option, especially on a device like a phone, is to limit the `depth` or `movetime` parameters in the `go` command. `movetime` is how much time to spend searching in milliseconds, once it's reached the engine will stop and announce its best move.
